app/controllers/application_controller.rb

```diff
@@ -1,3 +1,22 @@
 class ApplicationController < ActionController::Base
 
+  helper_method :current_user,
+                :logged_in?
+
+  private
+
+  def authenticate_user!
+    unless current_user
+      session[:required_page] = {controller: controller_name, action: action_name, id: params[:id]}
```

Comment:
```suggestion
      cookies[:required_page] = request.path
```
по ТЗ нужно использовать "куки", а не "сессию"
и наверное удобней использовать объект `request`, нежели конструировать хеш, как у тебя